### DISCUSSION TITLE
Should fonts scale in full-screen (i.e., presentation) mode?

### DISCUSSION THREAD

**Discussion by ooiikkjj** [kookma/TW-Tamasha#3] [Q&A]
issue: design looks great in browser window, but after using F11 to go full-screen, the fonts don't scale up...so in the presentation mode, especially when projected on the big screen the slides look empty.

**Comment by kookma** [comment]
@ooiikkjj -  the content size is set by width and height parameters. You can set them to like 90% and 90% to see how it works.
But the fonts do not change. I will add this features.

Note default width and height are 640px x 480px

**Reply by ooiikkjj** [reply]
e.g., this looks great in a window: 
![image](https://user-images.githubusercontent.com/75536247/105719669-8e731200-5ef0-11eb-89dc-3ff76af12592.png)

but in fullscreen, it doesn't: 
![image](https://user-images.githubusercontent.com/75536247/105719789-b5314880-5ef0-11eb-927c-91276421a9b5.png)

Thanks

**Reply by kookma** [reply]
Have you taken both screenshot on the same monitor (screen) or the second is from a projected slide on the wall or big screen like a 72'' monitor?

I use FF 84 on Windows 10 and both author mode and presentation (full screen = F11) show the same font size.  I have attached the screenshots in both modes.
[Tamasha-FF84.zip](https://github.com/kookma/TW-Tamasha/files/5867991/Tamasha-FF84.zip)

**Reply by ooiikkjj** [reply]
same screen, same monitor...first one was from a browser window which is about half the screen...second is from the same after using F11...they're the exact same font size...sorry if it looks like the font was shrinking.

the issue is that in the first one the slide looks good; same slide with full-screen looks empty because the font doesn't scale.

thanks for considering and looking at it.

**Comment by ooiikkjj** [comment]
also, how to sequence the slides in a Tamasha show?
(the old method of drag slides to reorder is not available now)
thanks

**Reply by kookma** [reply]
@ooiikkjj - I will add documentation. Tamasha uses the present methods. As it is work on tag, simple open one of the slides click on tag pill and reorder as you like. but I will add some simpler tool to do that.

**Reply by ooiikkjj** [reply]
thank you very much...that works fine...almost no need for a simpler tool

(sorry for my very basic questions...is a simpler way to center text, in the editor, instead of using CSS? does Shiraz do it?)

**Comment by silvyn** [comment]
Hi @kookma 

I thing you do something on the new beta release, but it create a side effect and hide text

I have a 1920 px laptop, and it happens in a lower resolution, see this gif test :

![tamasha](https://user-images.githubusercontent.com/9981592/106500062-abe04700-64c1-11eb-8c5d-61e17ccb778d.gif)

(With Firefox under Fedora Gnu/Linux, and also with Chromium)

**Reply by kookma** [reply]
@silvyn - would you please check the `$:/plugins/kookma/tamasha/styles/layout/content` and play with the setting and let me know the result.

Please note: the max height is 600px and font size scaled based on viewport width. The Long Long Tid is an example to show how long text is hidden because of overflow. So the recommendation here is to use small amount of text per slide.

**Reply by silvyn** [reply]
The text is hide an all tiddler demo, but if I understand it's logical if set to 600px (and full HD 1080px)

**Reply by silvyn** [reply]
I have set height to `auto`, and this do the job.

(also we need to change `max-width` to fit our need)

**Reply by kookma** [reply]
@silvyn thank you for your feedback. 

@NicolasPetton would you please have a look at this issue. One suggestion is

- use different styles for tamasha-slide-content in fullscreen and normal mode (I mean full screen presentation and author mode or not full screen)
- In full screen we can use both width and height as auto (this was your first suggestion for slide content)
- In normal mode (not full screen) height and width can be set
- The new problem with this solution is in non full screen mode we have to use the width and height of container to scale the font size. As these values are passed to presenter macro (or default value are used), I think we should be able to scale fonts without JS. Please let mw know your opinion

**Reply by NicolasPetton** [reply]
> * use different styles for tamasha-slide-content in fullscreen and normal mode (I mean full screen presentation and author mode or not full screen)
> 
> * In full screen we can use both width and height as auto (this was your first suggestion for slide content)

I think it makes sense.


> * In normal mode (not full screen) height and width can be set
> 
> * The new problem with this solution is in non full screen mode we have to use the width and height of container to scale the font size. As these values are passed to presenter macro (or default value are used), I think we should be able to scale fonts without JS. Please let mw know your opinion

I'm not sure how you want to scale the fonts based on the container without JavaScript. The `vw` unit is based on viewport width, so you'd have to rely on something else. What do you have in mind?

**Reply by kookma** [reply]
> I'm not sure how you want to scale the fonts based on the container without JavaScript. The `vw` unit is based on viewport width, so you'd have to rely on something else. What do you have in mind?

If the width/height are passed we can use them in a dynamic style sheet like

```
font-size: clac(1rem + $(cw)$*1.5);
```

The `cw` is the container width which is passed to presenter or a default value set in presenter header.


By the way if we have to use JS I am not against it.

**Reply by NicolasPetton** [reply]
> font-size: clac(1rem + $(cw)$*1.5);

But then you'd have to pass the dimensions of the slides as arguments to the macro, right? I liked the idea of having Tamasha figure out dimensions for the user.

Another approach which could fix both issues (the overflow on full screen and the dynamic font size in normal mode:

- use `calc(Xrem + Yvh)` in full screen, so that the font size depends on the the viewport **height**, not width
- use `calc(Xrem + $height$*Y)` in normal mode, assuming that `$height$` is known (like  600px today)

This way we can have dynamic font sizes in both modes, have slides take 600px height in normal mode, and set their height to auto in full-screen mode. 

**EDIT**: The trick would be to make font sizes scale in normal (600px height) and full-screen mode (dynamic height) to always have the same number of visible lines in a slide.

WDYT?

**Reply by kookma** [reply]
> > font-size: clac(1rem + $(cw)$*1.5);
> 
> But then you'd have to pass the dimensions of the slides as arguments to the macro, right? I liked the idea of having Tamasha figure out dimensions for the user.

That's right. But the dimensions are only used for slide-content in normal mode. Right now we have two container in presenter macro. an overlay container which works for full screen mode and author container which works for author mode. We can target tamasha-slide-content for both mode. The full screen mode uses auto height/weight or uses font size based on vw. The author mode uses the dimensions passed to the macro.


> 
> Another approach which could fix both issues (the overflow on full screen and the dynamic font size in normal mode:
> 
> * use `calc(Xrem + Yvh)` in full screen, so that the font size depends on the the viewport **height**, not width
> * use `calc(Xrem + $height$*Y)` in normal mode, assuming that `$height$` is known (like  600px today)
> 
> This way we can have dynamic font sizes in both modes, have slides take 600px height in normal mode, and set their height to auto in full-screen mode.
> 
> **EDIT**: The trick would be to make font sizes scale in normal (600px height) and full-screen mode (dynamic height) to always have the same number of visible lines in a slide.

This is a good solution and it will be wonderful to have no text reflow i.e. the same number of lines. Can we pass 600px height as macro input parameter or we have to use a fixed value?

> 
> WDYT?

**Reply by NicolasPetton** [reply]
> Can we pass 600px height as macro input parameter or we have to use a fixed value?

You could use `600px` as a default value and let the user override it, but do you think it would bring something to the user? Maybe it would, I don't know.

**Reply by kookma** [reply]
Okay, lets go with your solution.

**Comment by NicolasPetton** [comment]
@kookma I opened PR #12 which implements font sizes/line heights in presentation mode based on the viewport height.
